File: 13-Decorator/src/go/decorators.go
package main

import "fmt"

type MilkDecorator struct {
	wrapped Coffee
}

func WithMilk(c Coffee) Coffee {
	return &MilkDecorator{wrapped: c}
}

func (m *MilkDecorator) GetDescription() string {
	return fmt.Sprintf("%s, Milk", m.wrapped.GetDescription())
}

func (m *MilkDecorator) GetCost() float64 {
	return m.wrapped.GetCost() + 0.50
}

type WhipDecorator struct {
	wrapped Coffee
}

func WithWhip(c Coffee) Coffee {
	return &WhipDecorator{wrapped: c}
}

func (w *WhipDecorator) GetDescription() string {
	return fmt.Sprintf("%s, Whip", w.wrapped.GetDescription())
}

func (w *WhipDecorator) GetCost() float64 {
	return w.wrapped.GetCost() + 0.70
}

type CaramelDecorator struct {
	wrapped Coffee
}

func WithCaramel(c Coffee) Coffee {
	return &CaramelDecorator{wrapped: c}
}

func (c *CaramelDecorator) GetDescription() string {
	return fmt.Sprintf("%s, Caramel", c.wrapped.GetDescription())
}

func (c *CaramelDecorator) GetCost() float64 {
	return c.wrapped.GetCost() + 0.60
}

File: 13-Decorator/src/rust/decorators.rs
use crate::coffee::Coffee;

pub struct MilkDecorator {
    wrapped: Box<dyn Coffee>,
}

impl MilkDecorator {
    pub fn new(wrapped: Box<dyn Coffee>) -> Self {
        MilkDecorator { wrapped }
    }
}

impl Coffee for MilkDecorator {
    fn get_description(&self) -> String {
        format!("{}, Milk", self.wrapped.get_description())
    }
    fn get_cost(&self) -> f64 {
        self.wrapped.get_cost() + 0.50
    }
}

pub struct WhipDecorator {
    wrapped: Box<dyn Coffee>,
}

impl WhipDecorator {
    pub fn new(wrapped: Box<dyn Coffee>) -> Self {
        WhipDecorator { wrapped }
    }
}

impl Coffee for WhipDecorator {
    fn get_description(&self) -> String {
        format!("{}, Whip", self.wrapped.get_description())
    }
    fn get_cost(&self) -> f64 {
        self.wrapped.get_cost() + 0.70
    }
}

pub struct CaramelDecorator {
    wrapped: Box<dyn Coffee>,
}

impl CaramelDecorator {
    pub fn new(wrapped: Box<dyn Coffee>) -> Self {
        CaramelDecorator { wrapped }
    }
}

impl Coffee for CaramelDecorator {
    fn get_description(&self) -> String {
        format!("{}, Caramel", self.wrapped.get_description())
    }
    fn get_cost(&self) -> f64 {
        self.wrapped.get_cost() + 0.60
    }
}
